File: sample/src/main/java/pl/psi/aaas/sample/SimpleTestApp.java
package pl.psi.aaas.sample;

import pl.psi.aaas.*;
import pl.psi.aaas.engine.r.RConnectionProvider;
import pl.psi.aaas.engine.r.RServeEngine;
import pl.psi.aaas.usecase.CalculationDefinitionIf;
import pl.psi.aaas.usecase.CalculationDefinition;

import java.time.ZonedDateTime;
import java.util.HashMap;
import java.util.Map;

public class SimpleTestApp {
    public static void main(String[] args) {
        Facade< CalculationDefinitionIf, Void> facade = FixedTSFacade.INSTANCE;
        facade.callScript(prepCalcDef1());
    }

    private static CalculationDefinitionIf prepCalcDef1() {
        String[] strArr = {"a", "b", "c"};
        ZonedDateTime[] dtArr = {ZonedDateTime.now().minusHours(2), ZonedDateTime.now().minusHours(1), ZonedDateTime.now()};
        Vector<String> strVec = Parameter.ofArrayNotNull(strArr, String.class);
        Vector<ZonedDateTime> dtVec = Parameter.ofArrayNotNull(dtArr, ZonedDateTime.class);
        Vector<Long> longVec = Parameter.ofArrayNotNull(new Long[]{1L, 2L, 3L}, Long.class);
        Vector<Double> doubleVec = Parameter.ofArrayNotNull(new Double[]{0.1, 0.2, 1.0}, Double.class);
        Vector<Double> doubleNullVec = Parameter.ofArray(new Double[]{0.1, null, 1.0}, Double.class);
        Vector<Boolean> boolVec = Parameter.ofArrayNotNull(new Boolean[]{true, false, true}, Boolean.class);
        Vector<Boolean> boolNullVec = Parameter.ofArray(new Boolean[]{true, false, null}, Boolean.class);

        Column[] dfColumns = {
                new Column("dt", (Vector<Object>) (Vector<?>) dtVec),
                new Column("longs", (Vector<Object>) (Vector<?>) longVec),
                new Column("doubles", (Vector<Object>) (Vector<?>) doubleVec)
        };

        Map<String, Parameter<?>> parameters = new HashMap<>();
        parameters.put("str", Parameter.ofPrimitive("str_value"));
        parameters.put("dt", Parameter.ofPrimitive(ZonedDateTime.now()));
        parameters.put("d", Parameter.ofPrimitive(0.75));
        parameters.put("l", Parameter.ofPrimitive(10L));
        parameters.put("b", Parameter.ofPrimitive(false));
        parameters.put("strV", strVec);
        parameters.put("dtV", dtVec);
        parameters.put("longV", longVec);
        parameters.put("doubleV", doubleVec);
        parameters.put("doubleNullV", doubleNullVec);
        parameters.put("boolV", boolVec);
        parameters.put("boolNullV", boolNullVec);
        parameters.put("df", Parameter.ofDataFrame(dfColumns));

        return new CalculationDefinition(new HashMap<>(), new HashMap<>(), ZonedDateTime.now(), ZonedDateTime.now().plusDays(1), "add");
        // Note: CalculationDefinitionDTO in api module seems to have changed its constructor in my previous step.
        // I need to check the actual CalculationDefinitionDTO I wrote.
    }

    public static class LocalRConnectionProvider implements RConnectionProvider {
        @Override
        public String getHost() { return "192.168.99.100"; }
        @Override
        public int getPort() { return 6311; }
    }

    public static class FixedTSFacade implements Facade< CalculationDefinitionIf, Void> {
        public static final FixedTSFacade INSTANCE = new FixedTSFacade();
        private final Engine< CalculationDefinitionIf, Map<String, Parameter<?>>, Map<String, Parameter<?>>> engine = new RServeEngine(new LocalRConnectionProvider());

        @Override
        public Void callScript( CalculationDefinitionIf calcDef) {
            engine.call(calcDef);
            return null;
        }
    }
}

File: sample/src/main/kotlin/pl/psi/aaas/sample/SimpleTestApp.kt
package pl.psi.aaas.sample

import pl.psi.aaas.*
import pl.psi.aaas.engine.r.RConnectionProvider
import pl.psi.aaas.engine.r.REngineConfiguration
import pl.psi.aaas.engine.r.RServeEngine
import pl.psi.aaas.usecase.CalculationDefinition
import pl.psi.aaas.usecase.CalculationDefinitionDTO
import java.time.ZonedDateTime

object SimpleTestApp {
    @JvmStatic
    fun main(args: Array<String>) {
        val facade: Facade<CalculationDefinition, Unit> = FixedTSFacade
        facade.callScript(prepCalcDef1())
    }

    private fun prepCalcDef1(): CalculationDefinition {
        val strArr = arrayOf("a", "b", "c")
        val dtArr = arrayOf(ZonedDateTime.now().minusHours(2), ZonedDateTime.now().minusHours(1), ZonedDateTime.now())
        val strVec = Parameter.ofArrayNotNull(strArr, String::class.java)
        val dtVec = Parameter.ofArrayNotNull(dtArr, ZonedDateTime::class.java)
        val longVec = Parameter.ofArrayNotNull(arrayOf(1, 2, 3L), Long::class.java)
        val doubleVec = Parameter.ofArrayNotNull(arrayOf(0.1, 0.2, 1.0), Double::class.java)
        val doubleNullVec = Parameter.ofArray(arrayOf(0.1, null, 1.0), Double::class.java)
        val boolVec = Parameter.ofArrayNotNull(arrayOf(true, false, true), Boolean::class.java)
        val boolNullVec = Parameter.ofArray(arrayOf(true, false, null), Boolean::class.java)
        val dfColumns = arrayOf(Column("dt", dtVec as Vector<Any>), Column("longs", longVec as Vector<Any>), Column("doubles", doubleVec as Vector<Any>))

        val parameters = mutableMapOf(
                "str" to Parameter.ofPrimitive("str_value")
                , "dt" to Parameter.ofPrimitive(ZonedDateTime.now())
                , "d" to Parameter.ofPrimitive(0.75)
                , "l" to Parameter.ofPrimitive(10L)
                , "b" to Parameter.ofPrimitive(false)
                , "strV" to strVec
                , "dtV" to dtVec
                , "longV" to longVec
                , "doubleV" to doubleVec
                , "doubleNullV" to doubleNullVec
                , "boolV" to boolVec
                , "boolNullV" to boolNullVec
                , "df" to Parameter.ofDataFrame(dfColumns)
        )

        return CalculationDefinitionDTO("add",
                parameters,
                parameters)
    }
}

val localConfiguration = REngineConfiguration("192.168.99.100", 6311)

class LocalRConnectionProvider(override var configuration: REngineConfiguration = localConfiguration) : RConnectionProvider

object FixedTSFacade : Facade<CalculationDefinition, Unit> {
    private val engine: Engine<CalculationDefinition, Parameter<Any>, Parameter<Any>> = RServeEngine(LocalRConnectionProvider())

    override fun callScript(calcDef: CalculationDefinition) {
        engine.call(calcDef)
    }
}
